src/connections.jl

```diff
@@ -107,6 +107,24 @@ function handle_new_connection(jl_conn::Connection; throw_error::Bool=true)
         reset_encoding!(jl_conn)
     end
 
+    # keep a reference to `closed` so it's not cleaned up before the connection is
+    let closed = jl_conn.closed, conn_ptr = jl_conn.conn
```

Comment:
let block isn't required.
It will be captured by the closure.

Also that seems less clean than using the argument to the closure. (wich you have called `_`)

Reply:
You can't use the argument to the closure for `closed` because it may have been finalized beforehand if you don't keep a reference to it.

Reply:
Ok well if you close over it, it will hold a reference, anyway. so no need for the let-block.

I feel like though there have historically been a number of bugs in julia itself related to garbage collection/finalization order.
e.g.
https://github.com/JuliaLang/julia/issues/26939#issuecomment-385669964

Reply:
I don't know that closing over `jl_conn.closed` actually holds a reference to `closed` or just `jl_conn`.

Reply:
Verified that it does hold a reference to the connection:

```
julia> function baz()
           conn = LibPQ.Connection("dbname=postgres user=ericdavies")
           function bar()
               a = conn.closed
               a[]
           end
       end

julia> f = baz()

julia> @code_typed f()
CodeInfo(
1 ─ %1 = (Core.getfield)(#self#, :conn)::Connection
│   %2 = (Base.getfield)(%1, :closed)::Atomic{Bool}
│   %3 = $(Expr(:foreigncall, :(:jl_value_ptr), Ptr{Nothing}, svec(Any), :(:ccall), 1, :(%2)))::Ptr{Nothing}
│   %4 = (Base.bitcast)(Ptr{Bool}, %3)::Ptr{Bool}
│   %5 = (Base.Threads.llvmcall)("%ptr = inttoptr i64 %0 to i8*\n%rv = load atomic i8, i8* %ptr acquire, align 16\nret i8 %rv\n", Bool, Tuple{Ptr{Bool}}, %4)::Bool
└──      return %5
) => Bool
```

So the `let` is important for `closed`.

Reply:
Ok.
Why is this here, and not in the constructor?
I thought we were in the constructor, and thus have a reference to it from the constructor.

Reply:
`handle_new_connection` is responsible for checking if the connection is valid and closing it if it's not. Putting it in that function means we don't add the finalizer when we already know we have a bad connection and we're closing the connection anyway.

Reply:
Ok.